crypto/encode_decode/encoder_lib.c

```diff
@@ -312,120 +369,292 @@ OSSL_ENCODER_INSTANCE_get_output_type(OSSL_ENCODER_INSTANCE *encoder_inst)
     return encoder_inst->output_type;
 }
 
-static int encoder_process(OSSL_ENCODER_CTX *ctx, BIO *out)
+const char *
+OSSL_ENCODER_INSTANCE_get_output_structure(OSSL_ENCODER_INSTANCE *encoder_inst)
 {
-    size_t i, end;
-    void *latest_output = NULL;
-    size_t latest_output_length = 0;
-    const char *latest_output_type = NULL;
-    const char *last_input_type = NULL;
-    int ok = 0;
-
-    end = OSSL_ENCODER_CTX_get_num_encoders(ctx);
-    for (i = 0; i < end; i++) {
-        OSSL_ENCODER_INSTANCE *encoder_inst =
-            sk_OSSL_ENCODER_INSTANCE_value(ctx->encoder_insts, i);
-        OSSL_ENCODER *encoder = OSSL_ENCODER_INSTANCE_get_encoder(encoder_inst);
-        void *encoderctx = OSSL_ENCODER_INSTANCE_get_encoder_ctx(encoder_inst);
-        const char *current_input_type =
-            OSSL_ENCODER_INSTANCE_get_input_type(encoder_inst);
-        const char *current_output_type =
-            OSSL_ENCODER_INSTANCE_get_output_type(encoder_inst);
-        BIO *current_out;
-        BIO *allocated_out = NULL;
-        const void *current_data = NULL;
-        OSSL_PARAM abstract[3];
-        OSSL_PARAM *abstract_p;
-        const OSSL_PARAM *current_abstract = NULL;
-
-        if (latest_output_type == NULL) {
-            /*
-             * This is the first iteration, so we prepare the object to be
-             * encoded
-             */
+    if (encoder_inst == NULL)
+        return NULL;
+    return encoder_inst->output_structure;
+}
 
-            current_data = ctx->construct(encoder_inst, ctx->construct_data);
+static int encoder_process(struct encoder_process_data_st *data)
```

Comment:
This looks like an inefficient algorithm for finding a path on a graph.  Why not implement [Dijkstra's algorithm](https://en.wikipedia.org/wiki/Dijkstra%27s_algorithm)?  The recursion would be avoided, the computational complexity would be greatly reduced and the code would likely be easier to understand.

Reply:
Dijkstra's algorithm builds on existing knowledge, such as what nodes are neighbours and what the distance is between neighbours.  Where is that data?

Reply:
The output from an encoder has edges to the encoders that accept that format.  The graph can be built lazily: as new output formats are discovered, it can be populated on demand.  Although it would be better to create the graph and remember it, it isn't mandatory.

The distance for each edge will have to be unity, I don't think we can reasonably estimate the relative cost of each conversion.

Reply:
Hmmmm, are you talking about caching the graph of all encoders?  We could do that, of course, but that's beyond the scope of this PR.

Point is that essentially, `encoder_process` is figuring that graph from scratch...  and actually somewhat lazily (i.e. it stops as soon as the desired output format has been produced).

> The distance for each edge will have to be unity

...  and for such distances, we're basically talking about a true/false decision tree, yeah?  In that case, what's the practical difference between building the (decision) graph and applying Dijkstra's algorithm on it, and walking the stack in a trial and error fashion?

Reply:
Dijkstra's is more efficient?  It's systematic and will locate short paths preferentially.

Reply:
I'm not convinced that Dijkstra's is that efficient for all-unity edges when you're computing the graph on the fly... and the problem really seems to be about how to compute the graph efficiently and reliably.

A detail that you may be missing is that the current method uses the result of an encoding from each implementation to decide if there's an edge or not...  not just the pre-conditions (i.e. output type and structure must match between pairs of nodes), so the cost of figuring out another edge when you've already found any complete path to get the result is higher than you might expect.  Furthermore, the final node in that complete path writes the result directly to the BIO passed by the caller, so computing another, possibly successful path would have, er, surprising results.

I could see refactoring that so only the so called pre-conditions come into play to build the graph, and to only perform the actual encodings when you have that shortest path.  But then, if one of them fails, what do we do?  Choose an alternative path, possibly longer, or simply give up?

Also, unless I'm totally off my game, we must store away the (partial) graph somewhere.  An array in the heap, I presume?  I was trying to avoid that...  that's the primary reason I went with a recursive method.

Reply:
I agree with Richard that this request looks like a premature optimization. Of course it would be nice to actually build the full graph when new provider is loaded and enabled avoiding the need to always rebuild it or even build it incrementally on demand and cache it and then the Dijkstra's algo would be clear win. However I do not really see this caching/fast path finding being as necessary for 3.0 release.